File: pipeline/detectors/c_compilation_error.py
"""
Detector for C/C++ compilation errors when header files are missing.
"""

import re
import typing as T
from pipeline.detectors.base import Detector
from pipeline.models import ErrorClue


class CCompilationErrorDetector(Detector):
    """
    Detect C/C++ compilation errors when header files are missing.

    Matches patterns like:
    - /tmp/ex_bar.c:82:10: fatal error: ex.h: No such file or directory
    -    82 | #include "ex.h"
    - lib/src/node.c:2:10: fatal error: ./point.h: No such file or directory
    """

    PATTERNS = {
        "missing_file": r"fatal error:\s+\.?/?(?P<file_path>[^\s:]+):\s+No such file or directory",
    }

    EXAMPLES = [
        (
            "/tmp/ex_bar.c:82:10: fatal error: ex.h: No such file or directory",
            {
                "clue_type": "missing_file",
                "confidence": 1.0,
                "context": {
                    "file_path": "ex.h",
                },
            },
        ),
    ]

File: src/boil/detectors/c_compilation_error.go
// Detector for C/C++ compilation errors for missing files.
package detectors

// CCompilationErrorDetector detects C/C++ compilation errors for missing files
type CCompilationErrorDetector struct {
	*BaseDetector
}

func NewCCompilationErrorDetector() (*CCompilationErrorDetector, error) {
	base, err := NewBaseDetector(
		"CCompilationErrorDetector",
		100,
		map[string]string{
			"missing_file": `fatal error:\s+\.?/?(?P<file_path>[^\s:]+):\s+No such file or directory`,
		},
		[]DetectorExample{
			{
				Name:     "fatal_error_header",
				Input:    "/tmp/ex_bar.c:82:10: fatal error: ex.h: No such file or directory",
				ClueType: "missing_file",
				Context:  map[string]string{"file_path": "ex.h"},
			},
			{
				Name:     "fatal_error_relative_path",
				Input:    "lib/src/node.c:2:10: fatal error: ./point.h: No such file or directory",
				ClueType: "missing_file",
				Context:  map[string]string{"file_path": "point.h"},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &CCompilationErrorDetector{base}, nil
}
